modules/reload/src/main/scala/io/chrisdavenport/mules/reload/RefreshingCache.scala

```diff
@@ -0,0 +1,218 @@
+   **/
+  def size: F[Int] = innerCache.size
+
+  /**
+   * Insert an item in the cache, using the default expiration value of the cache.
+   */
+  def insert(k: K, v: V): F[Unit] =
+    innerCache.insert(k, v)
+
+  def insertWithTimeout(
+      optionTimeout: Option[TimeSpec]
+  )(k: K, v: V): F[Unit] =
+    innerCache.insertWithTimeout(optionTimeout)(k, v)
+
+  /**
+   * Return all keys present in the cache, including expired items.
+   **/
+  def keys: F[List[K]] =
+    innerCache.keys
+
+  def lookup(k: K): F[Option[V]] =
+    innerCache.lookup(k)
+
+  /**
+   * Try look up k. If the key is not present, will use fetch and insert using fetch
+   * This method always returns as is expected.
+   */
+  def lookupOrFetch(k: K)(fetch: F[V]): F[V] =
+    lookupOrFetch(defaultExpiration)(k)(fetch)
+
+  def lookupOrFetch(timeout: Option[TimeSpec])(k: K)(fetch: F[V]): F[V] =
+    lookup(k).flatMap {
+      case Some(v) => v.pure[F]
+      case None => fetchAndInsert(k, fetch, timeout)
+    }
+
+  def defaultExpiration: Option[TimeSpec] =
+    innerCache.defaultExpiration
+
+  /**
+   * Try look up k, if the key is not present it will setup an auto {@code refresh} every {@code period}
+   * This method is idempotent and always returns as expected.
+   * Note, by design, it does not override an existing refresh effect for the same k.
+   * To setup a new refresh method for it, you must
+   * first {@code delete(k)} and then call this method.
+   * If refresh keeps failing, the value will eventually expire and the refresh stops
+   */
+  def lookupOrRefresh(k: K, period: TimeSpec)(refresh: F[V]): F[V] =
+    lookupOrRefresh(k, period, defaultExpiration)(refresh)(PartialFunction.empty)
+
+  def lookupOrRefresh(k: K,
+                      period: TimeSpec,
+                      timeout: Option[TimeSpec])
+                     (refresh: F[V])
+                     (recoverError: PartialFunction[Throwable, F[Unit]]): F[V] =
+    lookup(k).flatMap {
+      case Some(v) => v.pure[F]
+      case None => setupRefresh(k,
+        refresh,
+        period,
+        timeout,
+        recoverError) *> fetchAndInsert(k, refresh, timeout)
+    }
+
+
+  private def setupRefresh(k: K,
+                           fetch: F[V],
+                           period: TimeSpec,
+                           timeout: Option[TimeSpec],
+                           recoverError: PartialFunction[Throwable, F[Unit]]): F[Unit] = {
+
+    def loop(): F[Unit] = {
+      def doRefreshFetch = {
+        lookup(k).flatMap[RefreshResult] {
+          case Some(_) => fetch.map[RefreshResult](Success(_))
+            .recoverWith(recoverError.andThen(_.as(ErrorRecovered))).recover {
+            case _ => ErrorUncovered
+          }
+          case None => NotFound.pure[F].widen
+        }
+      }
+
+      for {
+        fiber <- Concurrent[F].start(
+          Timer[F].sleep(Duration.fromNanos(period.nanos)) >>
+            doRefreshFetch
+        )
+        result <- fiber.join
```

Comment:
Why do we immediately join after starting?

Reply:
My bad, I kind of inherited this code from [AutoFetchingCache](https://github.com/ChristopherDavenport/mules/blob/master/modules/reload/src/main/scala/io/chrisdavenport/mules/reload/AutoFetchingCache.scala#L114-L119) 
Now you pointed it out I don't know if it's needed there. I will change this. 
Another thing I noticed in `AutoFetchingCache` is that all concurrency reads are going through guarded by the Semaphore, which might be a bottleneck.